localstack-core/localstack/runtime/runtime.py

```diff
@@ -0,0 +1,213 @@
+            self.components.gateway, self.config.GATEWAY_LISTEN, ssl_creds
+        )
+
+    def _on_starting(self):
+        self.starting.set()
+        hooks.on_runtime_start.run()
+
+    def _on_ready(self):
+        hooks.on_runtime_ready.run()
+        print(constants.READY_MARKER_OUTPUT, flush=True)
+        self.ready.set()
+
+    def _on_return(self):
+        LOG.debug("[shutdown] Running shutdown hooks ...")
+        hooks.on_runtime_shutdown.run()
+        LOG.debug("[shutdown] Cleaning up resources ...")
+        self._cleanup_resources()
+        self.stopped.set()
+        LOG.debug("[shutdown] Completed, bye!")
+
+    def _run_ready_monitor(self):
+        self._wait_for_gateway()
+        self._on_ready()
+
+    def _wait_for_gateway(self):
+        host_and_port = self.config.GATEWAY_LISTEN[0]
+
+        if not sync.poll_condition(
+            lambda: net.is_port_open(host_and_port.port), timeout=15, interval=0.3
+        ):
+            if LOG.isEnabledFor(logging.DEBUG):
+                # make another call with quiet=False to print detailed error logs
+                net.is_port_open(host_and_port.port, quiet=False)
+            raise TimeoutError(f"gave up waiting for gateway server to start on {host_and_port}")
+
+    def _clear_tmp_directory(self):
+        if self.config.CLEAR_TMP_FOLDER:
+            # try to clear temp dir on startup
+            try:
+                files.rm_rf(self.config.dirs.tmp)
+            except PermissionError as e:
+                LOG.error(
+                    "unable to delete temp folder %s: %s, please delete manually or you will "
+                    "keep seeing these errors.",
+                    self.config.dirs.tmp,
+                    e,
+                )
+
+    def _cleanup_resources(self):
+        threads.cleanup_threads_and_processes()
+        self._clear_tmp_directory()
+
+    # more legacy compatibility code
+    @property
+    def exit_code(self):
+        # FIXME: legacy compatibility code
+        from localstack.runtime import legacy
+
+        return legacy.EXIT_CODE.get()
+
+    @exit_code.setter
+    def exit_code(self, value):
+        # FIXME: legacy compatibility code
+        from localstack.runtime import legacy
+
+        legacy.EXIT_CODE.set(value)
+
+    def _run_shutdown_monitor(self):
+        # FIXME: legacy compatibility code. this can be removed once we replace access to the
+        #  ``SHUTDOWN_INFRA`` event with ``get_current_runtime().shutdown()``.
+        from localstack.runtime import legacy
+
+        legacy.SHUTDOWN_INFRA.wait()
+        self.shutdown()
+
+
+def create_from_environment() -> LocalstackRuntime:
+    """
+    Creates a new runtime instance from the current environment. It uses a plugin manager to resolve the
+    necessary components from the ``localstack.runtime.components`` plugin namespace to start the runtime.
+
+    TODO: perhaps we could control which components should be instantiated with a config variable/constant
```

Comment:
We should just directly map the environment variable to the name of the components plugin.
There's a few other implications there though that have been addressed in other code comments here like the need for a scoped config per component for example. (i.e. the config being part of the components plugin and thus only instantiated with them).